File: src/utils/quickReply.ts
/**
 * 快捷回复工具（纯函数）
 *
 * 负责 global + 角色级合并、排序、快捷键查找。
 * 宏展开在调用方（ChatInput）通过 expandMacros 完成。
 */

import type { QuickReply, QuickReplyStore } from '../../shared/types'

/** 获取角色有效快捷回复：global + 角色专属合并，按 order 排序 */
export function getEffectiveQuickReplies(store: QuickReplyStore, characterId: string | null | undefined): QuickReply[] {
  const global = store?.global ?? []
  const char = characterId ? (store?.byCharacter?.[characterId] ?? []) : []
  return [...global, ...char]
    .filter((q) => q.enabled)
    .sort((a, b) => a.order - b.order)
}

/** 按快捷键查找（1-9），无匹配返回 undefined */
export function findQuickReplyByHotkey(replies: QuickReply[], hotkey: number): QuickReply | undefined {
  return replies.find((q) => q.hotkey === hotkey)
}

/** 新建快捷回复 */
export function createQuickReply(partial: Partial<QuickReply> = {}): QuickReply {
  return {
    id: Math.random().toString(36).slice(2, 10),
    label: '快捷回复',
    content: '',
    action: 'text',
    sendWithAI: true,
    order: 0,
    enabled: true,
    ...partial,
  }
}

File: android/app/src/main/java/com/qingyu/companion/model/QuickReply.kt
package com.qingyu.companion.model

import kotlinx.serialization.Serializable

/**
 * 快捷回复。对齐 shared/types.ts 的 QuickReply。
 * 宏展开在 PC 侧完成（安卓端发的是用户消息，不是 prompt）。
 */
@Serializable
data class QuickReply(
    val id: String,
    val label: String,
    val content: String,
    /** text = 发送文本；preset = 切换预设；command = 触发斜杠命令 */
    val action: QuickReplyAction,
    val sendWithAI: Boolean,
    val order: Int,
    val enabled: Boolean,
    /** action=preset 时：目标预设 id */
    val presetId: String? = null,
    /** action=command 时：斜杠命令文本（如 /imagine） */
    val command: String? = null,
)

@Serializable
enum class QuickReplyAction {
    text,
    preset,
    command,
}

/** 快捷回复列表响应：全局 + 按角色专属（byCharacter 为角色 id → 列表） */
@Serializable
data class QuickReplyListResponse(
    val global: List<QuickReply>,
    val byCharacter: Map<String, List<QuickReply>> = emptyMap(),
)
